### ISSUE TITLE
argparse.add_mutually_exclusive_group fails for optional positional arguments

### ISSUE THREAD

**Issue by c103b3da-8d2d-4263-a996-57e557bc5081** [python/cpython#86020]
BPO | [41854](https://bugs.python.org/issue41854)
--- | :---

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2020-09-24.20:23:50.896>
labels = ['3.8', 'library']
title = 'argparse.add_mutually_exclusive_group fails for optional positional arguments'
updated_at = <Date 2020-10-17.07:41:14.402>
user = 'https://bugs.python.org/rrt'
```

bugs.python.org fields:
```python
activity = <Date 2020-10-17.07:41:14.402>
actor = 'rrt'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2020-09-24.20:23:50.896>
creator = 'rrt'
dependencies = []
files = []
hgrepos = []
issue_num = 41854
keywords = []
message_count = 3.0
messages = ['377460', '378795', '378803']
nosy_count = 2.0
nosy_names = ['paul.j3', 'rrt']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue41854'
versions = ['Python 3.8']
```

</p></details>

**Comment by allisonkarlitskaya** [comment]
There's another issue here, even with the workaround above: help output for positional parameters appearing as part of mutually exclusive groups is often broken.

Specifically: when a mutually exclusive group contains positional arguments and the usage line in the `--help` output is long enough that it needs to be wrapped, the group is no longer displayed as mutually exclusive.

Here's an example:

```python
#!/usr/bin/python3

import argparse

parser = argparse.ArgumentParser()
parser.add_argument('--a-flag')
parser.add_argument('--b-flag')
parser.add_argument('--c-flag')

group = parser.add_mutually_exclusive_group()
group.add_argument('--a-only')
group.add_argument('--b-only')
group.add_argument('positional', nargs='*', default=[])

args = parser.parse_args()
```

If you ask for `--help` with a wide enough terminal, you'll see correct output, with the "only" flags marked as mutually exclusive to the positional parameters:

```
usage: broken.py [-h] [--a-flag A_FLAG] [--b-flag B_FLAG] [--c-flag C_FLAG] [--a-only A_ONLY | --b-only B_ONLY | positional ...]
```

but if you make your terminal smaller, you get (incorrect):

```
usage: broken.py [-h] [--a-flag A_FLAG] [--b-flag B_FLAG] [--c-flag C_FLAG] [--a-only A_ONLY]
                 [--b-only B_ONLY]
                 [positional ...]
```

Removing the position parameters (ie: delete the `group.add_argument('positional', ...)`) gives the correct output.  Even in very small terminals (30 columns), it absolutely refuses to ever split the group onto multiple lines:

```
usage: broken.py [-h] [--a-flag A_FLAG]
                 [--b-flag B_FLAG]
                 [--c-flag C_FLAG]
                 [--a-only A_ONLY | --b-only B_ONLY]
```

This seems to be caused by the code in `_format_usage()` which attempts to consider `optionals` and `positionals` together in the unwrapped case:

```python
            # build full usage string
            format = self._format_actions_usage
            action_usage = format(optionals + positionals, groups)
            usage = ' '.join([s for s in [prog, action_usage] if s])
            pos_parts = _re.findall(part_regexp, pos_usage)
```

but considers them separately in case the output is too long:

```python
            # wrap the usage parts if it's too long
            text_width = self._width - self._current_indent
            if len(prefix) + len(usage) > text_width:
                ...
                opt_usage = format(optionals, groups)
                pos_usage = format(positionals, groups)
 ```

which is exacerbated by this code in `_format_actions_usage()` which assumes that the actions will be presented for formatting, complete, and in exactly the same order as they appear in the group:

```python
                start = actions.index(group._group_actions[0])
                ...
                end = start + len(group._group_actions)
                if actions[start:end] == group._group_actions:
```

Indeed, if you insert the positional parameters into group before the flag parameters, the grouping will always be broken, even if the terminal is infinitely wide:

```python
#!/usr/bin/python3

import argparse

parser = argparse.ArgumentParser()
group = parser.add_mutually_exclusive_group()
group.add_argument('positional', nargs='*', default=[])
group.add_argument('--flag')

args = parser.parse_args()
```

produces this incorrect output:

```
usage: broken.py [-h] [--flag FLAG] [positional ...]
```

**Comment by serhiy-storchaka** [comment]
The original issue was fixed (see #72795). Now positional arguments with `nargs='*'` are always considered optional, independently from the default value.

The bug reported in https://github.com/python/cpython/issues/86020#issuecomment-1114654477 is unrelated to this. It is either a duplicate of  #75949 or closely related.
